lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+        relation data, so we have to be careful about what our states mean.
+
+        Unit status can only be in "acquire", "release", "None" (None means unset)
+        Application status can only be in "granted" or "None" (None means unset or released)
+
+        """
+        unit_status = self.relation.data[self.unit].get('status', None)
+        app_status = self.relation.data[self.app].get(self.unit, self.IDLE)
+
+        if app_status == self.GRANTED and unit_status == self.RELEASE:
+            # Active release request.
+            return self.RELEASE
+
+        if app_status == self.IDLE and unit_status == self.ACQUIRE:
+            # Active acquire request.
+            return self.ACQUIRE
+
+        return app_status  # Granted or unset/released
+
+    @status.setter
+    def status(self, status):
+        """Set the given status.
+
+        Since we update the relation data, this may fire off a RelationChanged event.
+        """
+        if status == self.ACQUIRE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.GRANTED:
+            self.relation.data[self.app].update({self.unit: status})
+
+        if status is self.IDLE:
+            self.relation.data[self.app].update({self.unit: status})
+
+    def acquire(self):
+        self.status = self.ACQUIRE
+
+    def release(self):
+        self.status = self.RELEASE
+
+    def clear(self):
+        self.status = self.IDLE
+
+    def grant(self):
+        self.status = self.GRANTED
+
+    def is_held(self):
+        return self.status == self.GRANTED
+
+    def release_requested(self):
+        return self.status == self.RELEASE
+
+    def is_pending(self):
+        return self.status == self.ACQUIRE
+
+
+class Locks:
+    """Generator that returns a list of locks."""
+    def __init__(self, rel_name, charm):
+        self.rel_name = rel_name
+        self.charm = charm
+
+        # Gather all the units.
+        relation = charm.model.relations[rel_name][0]
+        units = [unit for unit in relation.units]
+
+        # Plus our unit ...
+        units.append(charm.framework.model.unit)
+
+        self.units = units
+
+    def __iter__(self):
+
+        for unit in self.units:
+            yield Lock(self.rel_name, self.charm, unit=unit)
+
+
+class RollingEvent(EventBase):
+    def __init__(self, handle, name):
+        super().__init__(handle)
+        self._name = name
+
+    def snapshot(self):
+        return {"name": self._name}
+
+    def restore(self, snapshot):
+        self._name = snapshot["name"]
+
+    @property
+    def name(self):
+        """The `name` property allows us to namespace Rolling Operations."""
+        return self._name
+
+
+class RunWithLock(RollingEvent):
+    pass
+
+
+class AcquireLock(RollingEvent):
+    pass
+
+
+class ProcessLocks(RollingEvent):
+    pass
+
+
+class RollingEvents(CharmEvents):
```

Comment:
Could there be merit in having a `RollingComplete` (name to be discussed...!) event that is emitted once *all* units have held the lock and run the callback? Such an event could be useful for charm authors to be able to subscribe to for clearing up statuses in the model?